### DISCUSSION TITLE
3D model not lit in scene

### DISCUSSION THREAD

**Discussion by keince** [pelicanmapping/osgearth#2393] [Support Q&A]
I have a need for rendering a 3D trail ribbon behind an aircraft model during flight somewhere on the Earth (mapNode is geocentric round Earth). I have successfully developed a 3D model that renders the ribbon using a triangle strip in localized world space, however, the model is not lit by the SkyNode like the 3D aircraft model that has the ribbon attached to it. 

I have verified that the geometry is correct along with normals bound per vertex and global color. On the trail itself, I have run

`ShaderGenerator gen;`
`gen.run(trail.get());`

Lighting is enabled on the trail node as

`trail->getOrCreateStateSet()->setMode(GL_LIGHTING, osg::StateAttribute::ON);`

Yet the trail model is still not lit. 

Do I need to write my own shader in order for lighting to work on the trail model?

**Comment by gwaldron** [comment]
Is your ribbon somewhere under the SkyNode in the scene graph?

(BTW, toggling lighting is shaderland is accomplished by calling `Lighting::set(stateset, osg::StateAttribute::ON)`. But this isn't likely your problem.)

**Reply by keince** [reply]
Yes, my scenegraph is laid out as follows...

root->skyNode->mapNode->trailGroup->trail

I will also try the lighting toggle suggestion to see if that makes any difference.

**Comment by keince** [comment]
I changed to `Lighting::set(stateset, osg::StateAttribute::ON)` which had no effect. The trail model is still unlit.

Here is my code for setting up the trail node...

```
osg::ref_ptr<osg::Vec3Array> vertices = new osg::Vec3Array();
osg::ref_ptr<osg::Vec3Array> normals = new osg::Vec3Array();
osg::ref_ptr<osg::Vec4Array> colors = new osg::Vec4Array();
colors->push_back(osg::Vec4(1.0f, 0.0f, 0.0f, 1.0f));

trail = new osg::Geometry();
trail->setDataVariance(osg::Object::DYNAMIC);
trail->setUseDisplayList(false);
trail->setUseVertexBufferObjects(true);

Lighting::set(trail->getOrCreateStateSet(), osg::StateAtribute::ON);
drawArrays = new osg::DrawArrays(GL_TRIANGLE_STRIP);
trail->addPrimitiveSet(drawArrays);
trail->setVertexArray(vertices.get());
trail->setNormalArray(normals.get(), osg::Array::BIND_PER_VERTEX);
trail->setColorArray(colors.get(), osg::Array::BIND_OVERALL);

ShaderGenerator gen;
gen.run(trail.get());

addChild(trail);
```

The vertices and normals of the trail are added and removed dynamically as the aircraft flies along.

Perhaps I'm doing something out of order or incorrect above? As I mentioned before, the trail draws and looks correct except for it not being lit.

**Comment by gwaldron** [comment]
Honestly you shouldn't need to call Lighting or even generate a shader.

Do me a favor and try using the "GL" sky; you can activate this with the `--sky-low` command line with osgearth_viewer or osgearth_imgui.
